samples/rust/rust_seq_file.rs

```diff
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Example of using a [`seq_file`] in Rust.
+//!
+//! C header: [`include/linux/seq_file.h`](../../../include/linux/seq_file.h)
+//! C header: [`include/linux/proc_fs.h`](../../../include/linux/proc_fs.h)
+//!
+//! Reference: <https://www.kernel.org/doc/html/latest/filesystems/seq_file.html>
+
+#![no_std]
+#![feature(allocator_api, global_asm, try_reserve)]
+
+use alloc::{boxed::Box, sync::Arc};
+use core::{
+    convert::TryInto,
+    fmt::Write,
+    iter::{repeat, Peekable, Repeat, Take},
+    pin::Pin,
+};
+use kernel::{
+    c_str,
+    file::File,
+    file_operations::{FileOpener, FileOperations},
+    io_buffer::IoBufferWriter,
+    miscdev, mutex_init,
+    prelude::*,
+    proc_fs, seq_file,
+    sync::Mutex,
+    Error, Result,
+};
+
+module! {
+    type: RustSeqFileDev,
+    name: b"rust_seq_file",
+    author: b"Adam Bratschi-Kaye",
+    description: b"Rust sample using a seq_file",
+    license: b"GPL v2",
+}
+
+struct State(Mutex<u32>);
+
+impl seq_file::SeqOperations for State {
+    type Item = String;
+    type Iterator = Take<Repeat<String>>;
+    type DataWrapper = Arc<State>;
+    type IteratorWrapper = Box<Peekable<Self::Iterator>>;
+
+    fn display(item: &Self::Item) -> &str {
+        &item[..]
+    }
+
+    fn start(&self) -> Result<Self::IteratorWrapper> {
+        const MAX_DIGITS: usize = 3;
+        const MAX_LENGTH: usize = MAX_DIGITS + 1;
+        const MAX_COUNT: u32 = 10u32.pow(MAX_DIGITS as u32);
+
+        let count = self.0.lock();
+        let mut message = String::new();
+
+        let template = "rust_seq_file: device opened this many times: ";
+        message.try_reserve_exact(template.len() + MAX_LENGTH)?;
+        // NOPANIC: We reserved space for `template` above.
+        message.push_str(template);
+        if *count < MAX_COUNT {
+            // NOPANIC: There are MAX_LENGTH characters remaining in the string which
+            // leaves space for a MAX_DIGITS digit number and the newline.
+            writeln!(&mut message, "{}", *count).map_err(|_| Error::ENOMEM)?;
+        }
+
+        Box::try_new(repeat(message).take((*count).try_into()?).peekable())
+            .map_err(|_| Error::ENOMEM)
+    }
+}
+
+#[derive(Clone)]
+struct SharedState(Arc<State>);
+
+impl SharedState {
+    fn try_new() -> Result<Self> {
+        let state = Arc::try_new(
+            // SAFETY: `mutex_init!` is called below.
+            unsafe { State(Mutex::new(0)) },
+        )?;
+        // SAFETY: Mutex is pinned behind `Arc`.
+        let pin_state = unsafe { Pin::new_unchecked(&state.0) };
+        mutex_init!(pin_state, "SharedState::0");
+        Ok(SharedState(state))
+    }
+}
+
+impl FileOpener<SharedState> for SharedState {
+    fn open(ctx: &SharedState) -> Result<Self::Wrapper> {
+        pr_info!("rust seq_file was opened!\n");
+        Ok(Box::try_new(ctx.clone())?)
+    }
+}
+
+impl FileOperations for SharedState {
+    type Wrapper = Box<Self>;
```

Comment:
nit: this is the default type, so it can omitted.

Reply:
:+1: